dusty/systems/nfs/server.py

```diff
@@ -0,0 +1,91 @@
+
+import logging
+import os
+from subprocess import CalledProcessError
+
+from ... import constants
+from .. import config_file
+from ...compiler.spec_assembler import get_all_repos
+from ...log import log_to_client
+from ...source import Repo
+from ..virtualbox import get_docker_vm_ip
+from ...subprocess import check_and_log_output_and_error, check_output
+
+def configure_nfs_server():
+    """
+    This function is used with `dusty up`.  It will check all active repos to see if
+    they are exported.  If any are missing, it will replace current dusty exports with
+    exports that are needed for currently active repos, and restart
+    the nfs server
+    """
+    vm_ip = get_docker_vm_ip()
+    repos_for_export = get_all_repos(active_only=True, include_specs_repo=False)
+
+    current_exports = _exports_set()
+    needed_exports = _get_exports_for_repos(repos_for_export, vm_ip)
+
+    if not needed_exports.difference(current_exports):
+        # All needed exports are already satisfied in /etc/exports
+        return
+
+    _write_exports_config(needed_exports)
+    _restart_server()
+
+def add_exports_for_repos(repos):
+    """
+    This function will add needed entries to /etc/exports.  It will not remove any
+    entries from the file.  It will then restart the server if necessary
+    """
+    vm_ip = get_docker_vm_ip()
+    current_exports = _exports_set()
+    needed_exports = _get_exports_for_repos(repos, vm_ip)
+
+    if not needed_exports.difference(current_exports):
+        # All needed exports are already satisfied in /etc/exports
+        return
+
+    _write_exports_config(current_exports.union(needed_exports))
+    _restart_server()
+
+def _get_exports_for_repos(repos, vm_ip):
+    config_set = set([_export_for_dusty_managed(vm_ip)])
+    for repo in repos:
+        if not repo.is_overridden:
+            continue
+        config_set.add(_export_for_repo(repo, vm_ip))
+    return config_set
+
+def _write_exports_config(exports_set):
+    exports_config = ''.join(exports_set)
+    if os.path.isfile(constants.EXPORTS_PATH):
+        current_config = config_file.read(constants.EXPORTS_PATH)
+    else:
+        current_config = ''
+    current_config = config_file.remove_current_dusty_config(current_config)
+    current_config += config_file.create_config_section(exports_config)
+    config_file.write(constants.EXPORTS_PATH, current_config)
+
+def _export_for_dusty_managed(vm_ip):
+    return '{} {} -alldirs -maproot=0:0\n'.format(os.path.realpath(constants.REPOS_DIR), vm_ip)
+
+def _export_for_repo(repo, vm_ip):
+    return '{} {} -alldirs -maproot={}\n'.format(os.path.realpath(repo.local_path), vm_ip, _maproot_for_repo(repo))
+
+def _maproot_for_repo(repo):
+    stat = os.stat(repo.local_path)
+    return '{}:{}'.format(stat.st_uid, stat.st_gid)
+
+def _restart_server():
+    try:
+        check_output(['nfsd', 'update'], demote=False)
+    except CalledProcessError as e:
+        if 'nfsd not running' in e.output:
```

Comment:
It's more direct and less brittle than checking the output for a particular string.